### PR TITLE
Add Unsqueeze and more support for squeeze 

### PR DESCRIPTION
**Pull request by ntarafdar**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/12731)

### Problem description
There's no Unsqueeze support in ttnn. there's also squeeze but only in python with limited support

### What's changed
Added Unsqueeze and squeeze as wrappers to reshape

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/test_squeeze.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import torch
+
+import ttnn
+
+
+@pytest.mark.parametrize(
+    "input_shape, dim, output_shape",
+    [
+        # ((1,256,1), -1, (1,256)),
+        ((1, 1, 1, 256), 2, (1, 1, 256)),
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/test_squeeze.py:16]: Torch supports `dim` argument to specify a dimension or a list of dimensions. Can we reflect that in a test and indicate that only 0 is supported if thats the case?

File: tests/ttnn/unit_tests/test_unsqueeze.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import torch
+
+import ttnn
+
+
+@pytest.mark.parametrize(
+    "input_shape, dim, output_shape",
+    [
+        ((1, 1, 256), 2, (1, 1, 1, 256)),
```
Comment (dmakoviichuk-tt) [inline tests/ttnn/unit_tests/test_unsqueeze.py:15]: can we use non divisible by 32 shapes in last dimenstions too?

File: tests/ttnn/unit_tests/test_unsqueeze.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        ((1, 256), 1, (1, 1, 256)),
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/test_unsqueeze.py:16]: can you add a case with a negative dim value?

File: tests/ttnn/unit_tests/test_squeeze.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import torch
+
+import ttnn
+
+
+@pytest.mark.parametrize(
+    "input_shape, dim, output_shape",
+    [
+        # ((1,256,1), -1, (1,256)),
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/test_squeeze.py:15]: tracked by a separate ticket?

File: ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+
+#include "squeeze.hpp"
+#include "ttnn/operations/core/core.hpp"
+
+namespace ttnn::operations::data_movement {
+
+ttnn::Tensor SqueezeOperation::invoke(
+    const ttnn::Tensor& input_tensor,
+    const int dim
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp:13]: makes sense to check if shape[dim] == 1, right?

File: tests/ttnn/unit_tests/test_squeeze.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/test_squeeze.py:1]: 2024

File: tests/ttnn/unit_tests/test_unsqueeze.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/test_unsqueeze.py:1]: 2024

File: ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze_pybind.hpp [hunk 8]
```diff
@@ -0,0 +1,13 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze_pybind.hpp:1]: 2024

File: ttnn/cpp/ttnn/operations/data_movement/unsqueeze/unsqueeze_pybind.cpp [hunk 9]
```diff
@@ -0,0 +1,64 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/unsqueeze/unsqueeze_pybind.cpp:1]: 2024

File: ttnn/cpp/ttnn/operations/data_movement/unsqueeze/unsqueeze_pybind.hpp [hunk 10]
```diff
@@ -0,0 +1,13 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/unsqueeze/unsqueeze_pybind.hpp:1]: 2024

File: tests/ttnn/unit_tests/test_squeeze.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    [
+        # ((1,256,1), -1, (1,256)),
+        ((1, 1, 1, 256), 2, (1, 1, 256)),
+    ],
+)
+def test_squeeze_as_reshape(device, input_shape, dim, output_shape):
+    torch_input_tensor = torch.rand(input_shape, dtype=torch.float32)
+    torch_squeeze_tensor = torch.squeeze(torch_input_tensor, dim)
+    print(torch_input_tensor.shape)
+    print(torch_squeeze_tensor.shape)
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/test_squeeze.py:23]: Remove?

**Review by TT-BrianLiu — APPROVED** [review]
Minor comments

**Review by sjameelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze_pybind.cpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        doc,
+        ttnn::pybind_overload_t{
+            [](const data_movement_operation_t& self,
+               const ttnn::Tensor& input_tensor,
+               const int dim
+             ) -> ttnn::Tensor {
+                return self(input_tensor, dim);
+            },
+            py::arg("input_tensor"),
+            py::arg("dim")
```
[the remark this answers is not in the document]
Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze_pybind.cpp:32]: Ya for this PR, going to only support value as that's what the use case in Bert. I'll add that in the documentation for the op.

File: tests/ttnn/unit_tests/test_squeeze.py (hunk 4 above)
Comment (tarafdarTT) [reply tests/ttnn/unit_tests/test_squeeze.py]: This one is a reshape issue. We can discuss this. Unsqueeze prepares the new shape for reshape, and here reshape  complains 
`Unable to reshape given tensor!` Tracked in https://github.com/tenstorrent/tt-metal/issues/12715
Need to get this staffed asap, will discuss with @eyonland . I can work on it as well.

File: ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp (hunk 5 above)
Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp:13]: in PyTorch if its not 1, it is ignored, I'll add that to the documentation

File: ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            output_shape_vector.push_back(tensor_shape[i]);
+        }
+    }
+
+    // If dim is out of range or original dimension was not of size 1, include all dimensions
+    if (dim >= static_cast<int>(tensor_shape.size()) || tensor_shape[dim] != 1) {
+        return input_tensor;
+    }
+
+    ttnn::Shape output_shape(output_shape_vector);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp:38]: we just lost any padding info here, right?

Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp:38]: yes thats true, the test cases for BERT I have are Row Major which have no padding.

As a first stab to get Bert unblocked I'm going to add RM support and Tile support with no padding. We can revisit this op to expand it.

Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp:38]: I'll add a TT_FATAL in the op to enforce this constraint

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/data_movement/squeeze/squeeze.cpp:38]: Thank you

**Review by ayerofieiev-tt — APPROVED** [review]
